File: go/sorting.go
package main

import "container/heap"
import "sort"

type Pair struct {
	A int
	B int
}

// return the index of first element which is out of order
// return len(v) if v is sorted
func IsSortedUntil(v []int, cp func(int, int) bool) int {
	size := len(v)
	for i := 1; i < size; i++ {
		if !cp(v[i-1], v[i]) {
			return i
		}
	}
	return size
}

func IsSorted(v []int, cp func(int, int) bool) bool {
	return IsSortedUntil(v, cp) == len(v)
}

// quicksort
func Sort(v []int, cp func(int, int) bool) {
	size := len(v)
	if size <= 1 {
		return
	}
	pivot := size / 2
	x := v[pivot]
	v[pivot], v[size-1] = v[size-1], v[pivot]
	j := 0
	for i := 0; i < size-1; i++ {
		if cp(v[i], x) {
			v[i], v[j] = v[j], v[i]
			j++
		}
	}
	v[j], v[size-1] = v[size-1], v[j]
	Sort(v[:j], cp)
	Sort(v[j+1:], cp)
}

// insert sort
func StableSort(v []Pair, cp func(*Pair, *Pair) bool) {
	size := len(v)
	if size > 1 {
		for i := 1; i < size; i++ {
			j := i
			e := v[j]
			for k := i; k > 0 && cp(&e, &v[k-1]); j-- {
				k--
				v[j] = v[k]
			}
			v[j] = e
		}
	}
}

func PartialSort(in []int, border int) {
	size := len(in)
	h := IntHeap(in[:border])
	heap.Init(&h)
	for i := border; i < size; i++ {
		if in[i] < h[0] {
			in[i], h[0] = h[0], in[i]
			x := heap.Pop(&h)
			h = append(h, x.(int))
		}
	}
	SortHeap(&h)
}

func PartialSortCopy(v []int, border int, out []int) {
	in := v[:border]
	inSize := len(in)
	outSize := len(out)
	if inSize <= outSize {
		copy(out, in)
		sort.Ints(out[:inSize])
	} else {
		space := make([]int, inSize)
		copy(space, in)
		sort.Ints(space)
		copy(out, space)
	}
}

func quickSelect(v []int, left int, right int, border int) int {
	v[border], v[right] = v[right], v[border]
	storeIdx := left
	for i := left; i < right; i++ {
		if v[i] < v[right] {
			v[storeIdx], v[i] = v[i], v[storeIdx]
			storeIdx++
		}
	}
	v[right], v[storeIdx] = v[storeIdx], v[right]
	return storeIdx
}

//The difference between PartialSort and NthElement is PartialSort sorts the
// elements in [0, border), while NthElement just makes sure that elements in [0, border) are all less than element in border
func NthElement(v []int, border int) {
	left := 0
	right := len(v) - 1
	for {
		if left >= border || right <= border {
			break
		}
		nth := quickSelect(v, left, right, border)
		if nth == border { // left <= nth, border <= right
			break
		}
		if nth < border {
			left = nth
		} else {
			right = nth
		}
	}
}

File: ruby/sorting.rb
# Implementations of sorting algorithms.

require_relative 'heap'

module Sorting
include Heap

def is_sorted(seq)
  (0...(seq.size-1)).each{|i| return false if seq[i+1] < seq[i]}
  true
end

def is_sorted_until(seq)
  (0...(seq.size-1)).each{|i| return i+1 if seq[i+1] < seq[i]}
  seq.size
end

def sort(seq)
  def quick_sort(seq, start, stop)
    return if start >= stop
    i = start - 1
    (start...stop).each do |j|
      if seq[j] <= seq[stop]
        i += 1
        seq[i], seq[j] = seq[j], seq[i]
      end
    end
    seq[i+1], seq[stop] = seq[stop], seq[i+1]
    quick_sort(seq, start, i)
    quick_sort(seq, i + 2, stop)
  end

  quick_sort(seq, 0, seq.size-1)
end

def partial_sort(seq, border)
  make_heap seq
  Array.new(border){pop_heap(seq)}.reverse_each{|v| seq.unshift(v)}
end

def partial_sort_copy(seq, border)
  make_heap seq
  Array.new(border){pop_heap(seq)}
end

# use insert sort to make it stable
# accept a block as less comparator
def stable_sort(seq)
  (1...seq.size).each do |i|
    j = i
    while j > 0 && (yield seq[j], seq[j-1])
      seq[j-1], seq[j] = seq[j], seq[j-1]
      j -= 1
    end
  end
end

def nth_element(seq, nth)
  def quick_select(seq, left, right, border)
    return if left >= right
    i = left - 1
    (left...right).each do |j|
      if seq[j] <= seq[right]
        i += 1
        seq[i], seq[j] = seq[j], seq[i]
      end
    end
    seq[i+1], seq[right] = seq[right], seq[i+1]
    if i + 1 < border
      quick_select(seq, i+2, right, border)
    elsif i + 1 > border
      quick_select(seq, left, i, border)
    end
  end

  quick_select(seq, 0, seq.size-1, nth)
  seq[nth]
end

end
